> Thanks. A bit surprised you found a solution that easily since from what I remember it was super hard to get #37543 working, but 🤷

Maybe this comment confused everyone https://github.com/quarkusio/quarkus/issues/36376#issuecomment-1755639759? I guess it would be a correct assumption for most things, but SmallRye Config is part of the core dependencies, so you can directly reference the exception.

> Any chance you can put the revert commit and your alternative solution in separate commits? I'm having trouble understanding what's what :/

Sure. Done!

> Also since there's no automated test... we'll need to check what happens with the original reproduceer from #36376 . I'll do that later, as that involves switching branches on my local repo and rebuilding Quarkus and well... that takes time and I'm currently working on something else.

I did try it with your reproducer. This is how it looks like:

```console
[INFO] Running org.acme.GreetingResourceIT
2024-06-13 11:09:46,315 INFO  [org.tes.doc.DockerClientProviderStrategy] (build-14) Loaded org.testcontainers.dockerclient.UnixSocketClientProviderStrategy from ~/.testcontainers.properties, will try it first
2024-06-13 11:09:46,710 INFO  [org.tes.doc.DockerClientProviderStrategy] (build-14) Found Docker environment with local Unix socket (unix:///var/run/docker.sock)
2024-06-13 11:09:46,712 INFO  [org.tes.DockerClientFactory] (build-14) Docker host IP address is localhost
2024-06-13 11:09:46,740 INFO  [org.tes.DockerClientFactory] (build-14) Connected to docker:
  Server Version: 24.0.2
  API Version: 1.43
  Operating System: Docker Desktop
  Total Memory: 16010 MB
2024-06-13 11:09:46,753 INFO  [org.tes.ima.PullPolicy] (build-14) Image pull policy will be performed by: DefaultPullPolicy()
2024-06-13 11:09:46,757 INFO  [org.tes.uti.ImageNameSubstitutor] (build-14) Image name substitution will be performed by: DefaultImageNameSubstitutor (composite of 'ConfigurationFileImageNameSubstitutor' and 'PrefixingImageNameSubstitutor')
2024-06-13 11:09:46,762 INFO  [org.tes.DockerClientFactory] (build-14) Checking the system...
2024-06-13 11:09:46,763 INFO  [org.tes.DockerClientFactory] (build-14) ✔︎ Docker server version should be at least 1.6.0
2024-06-13 11:09:46,830 INFO  [tc.doc.io/postgres:14] (build-14) Creating container for image: docker.io/postgres:14
2024-06-13 11:09:47,316 INFO  [org.tes.uti.RegistryAuthLocator] (build-14) Credential helper/store (docker-credential-desktop) does not have credentials for docker.io
2024-06-13 11:09:47,445 INFO  [tc.doc.io/postgres:14] (build-14) Reusing container with ID: 57be00faba6e1fecd75fcc0be76743532d8e45c2829254dad94d7fcb60ef4d1d and hash: f915c881d2926ee5f0d87578e2cbd39fef89f562
2024-06-13 11:09:47,446 INFO  [tc.doc.io/postgres:14] (build-14) Reusing existing container (57be00faba6e1fecd75fcc0be76743532d8e45c2829254dad94d7fcb60ef4d1d) and not creating a new one
2024-06-13 11:09:47,687 INFO  [tc.doc.io/postgres:14] (build-14) Container docker.io/postgres:14 started in PT0.856765S
2024-06-13 11:09:47,688 INFO  [tc.doc.io/postgres:14] (build-14) Container is started (JDBC URL: jdbc:postgresql://localhost:58573/quarkus?loggerLevel=OFF)
2024-06-13 11:09:47,688 INFO  [io.qua.dev.pos.dep.PostgresqlDevServicesProcessor] (build-14) Dev Services for PostgreSQL started.
2024-06-13 11:09:47,689 INFO  [io.qua.dat.dep.dev.DevServicesDatasourceProcessor] (build-14) Dev Services for default datasource (postgresql) started - container ID is 57be00faba6e
2024-06-13 11:09:47,692 INFO  [io.qua.hib.orm.dep.dev.HibernateOrmDevServicesProcessor] (build-7) Setting quarkus.hibernate-orm.database.generation=drop-and-create to initialize Dev Services managed database
Executing "/Users/radcortez/.sdkman/candidates/java/17.0.5-tem/bin/java -Dquarkus.http.port=8081 -Dquarkus.http.ssl-port=8444 -Dtest.url=http://localhost:8081 -Dquarkus.log.file.path=/Users/radcortez/Code/open-source/quarkus-reproducers/quarkus-36376/target/quarkus.log -Dquarkus.log.file.enable=true -Dquarkus.log.category."io.quarkus".level=INFO -Dquarkus.profile=integrationtest -Dquarkus.datasource.password=quarkus -Dquarkus.datasource.devservices.reuse=true -Dquarkus.datasource.db-kind=postgresql -Dquarkus.datasource.jdbc.url=jdbc:postgresql://localhost:58573/quarkus?loggerLevel=OFF -Dquarkus.datasource.username=quarkus -Dquarkus.hibernate-orm.database.generation=drop-and-create -jar /Users/radcortez/Code/open-source/quarkus-reproducers/quarkus-36376/target/quarkus-app/quarkus-run.jar"
Jun 13, 2024 11:09:49 AM org.hibernate.Version
INFO: HHH000412: Hibernate ORM core version 6.0.6.Final
Configuration validation failed:
	java.util.NoSuchElementException: SRCFG00011: Could not expand value some-unresolved-property in property hello.message
Log file '/Users/radcortez/Code/open-source/quarkus-reproducers/quarkus-36376/target/quarkus.log' was not created. Check if the options quarkus.log.level and quarkus.log.file.level are at least INFO (or more verbose).
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 65.93 s <<< FAILURE! - in org.acme.GreetingResourceIT
[ERROR] org.acme.GreetingResourceIT.testHelloEndpoint  Time elapsed: 0.013 s  <<< ERROR!
java.lang.RuntimeException: java.lang.IllegalStateException: Unable to determine the status of the running process. See the above logs for details
	at io.quarkus.test.junit.QuarkusIntegrationTestExtension.throwBootFailureException(QuarkusIntegrationTestExtension.java:373)
	at io.quarkus.test.junit.QuarkusIntegrationTestExtension.beforeEach(QuarkusIntegrationTestExtension.java:117)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: java.lang.IllegalStateException: Unable to determine the status of the running process. See the above logs for details
	at io.quarkus.test.common.LauncherUtil.waitForCapturedListeningData(LauncherUtil.java:102)
	at io.quarkus.test.common.DefaultJarLauncher.start(DefaultJarLauncher.java:72)
	at io.quarkus.test.junit.IntegrationTestUtil.startLauncher(IntegrationTestUtil.java:194)
	at io.quarkus.test.junit.QuarkusIntegrationTestExtension.doProcessStart(QuarkusIntegrationTestExtension.java:301)
	at io.quarkus.test.junit.QuarkusIntegrationTestExtension.ensureStarted(QuarkusIntegrationTestExtension.java:169)
	at io.quarkus.test.junit.QuarkusIntegrationTestExtension.beforeAll(QuarkusIntegrationTestExtension.java:130)
	... 1 more
```

When running directly (with another example):

```console
java -jar target/quarkus-app/quarkus-run.jar
Configuration validation failed:
	java.util.NoSuchElementException: SRCFG00011: Could not expand value missing in property my-config.expand
	SRCFG00050: my-config.unmapped in PropertiesConfigSource[source=jar:file:///Users/radcortez/Code/open-source/quarkus-reproducers/quarkus-39090/target/quarkus-app/app/quarkus-issue-39090-1.0.0-SNAPSHOT.jar!/application.properties] does not map to any root
	SRCFG00050: my-config.another in PropertiesConfigSource[source=jar:file:///Users/radcortez/Code/open-source/quarkus-reproducers/quarkus-39090/target/quarkus-app/app/quarkus-issue-39090-1.0.0-SNAPSHOT.jar!/application.properties] does not map to any root
```

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/RuntimeConfigSetupBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/RuntimeConfigSetupBuildStep.java
@@ -10,14 +10,11 @@ import io.quarkus.deployment.annotations.Produce;
 import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
 import io.quarkus.deployment.builditem.MainBytecodeRecorderBuildItem;
 import io.quarkus.deployment.builditem.RuntimeConfigSetupCompleteBuildItem;
-import io.quarkus.gizmo.CatchBlockCreator;
 import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.ClassOutput;
 import io.quarkus.gizmo.MethodCreator;
-import io.quarkus.gizmo.TryBlock;
 import io.quarkus.runtime.StartupContext;
 import io.quarkus.runtime.StartupTask;
-import io.quarkus.runtime.configuration.ConfigurationException;
 
 public class RuntimeConfigSetupBuildStep {
     private static final String RUNTIME_CONFIG_STARTUP_TASK_CLASS_NAME = "io.quarkus.deployment.steps.RuntimeConfigSetup";
@@ -40,17 +37,11 @@ public class RuntimeConfigSetupBuildStep {
                 .interfaces(StartupTask.class).build()) {
 
             try (MethodCreator method = clazz.getMethodCreator("deploy", void.class, StartupContext.class)) {
-                TryBlock tryBlock = method.tryBlock();
-                tryBlock.invokeVirtualMethod(
-                        ofMethod(StartupContext.class, "setCurrentBuildStepName", void.class, String.class),
+                method.invokeVirtualMethod(ofMethod(StartupContext.class, "setCurrentBuildStepName", void.class, String.class),
                         method.getMethodParam(0), method.load("RuntimeConfigSetupBuildStep.setupRuntimeConfig"));
 
-                tryBlock.invokeStaticMethod(C_CREATE_RUN_TIME_CONFIG);
-                tryBlock.returnValue(null);
-
-                CatchBlockCreator cb = tryBlock.addCatch(RuntimeException.class);
-                cb.throwException(ConfigurationException.class, "Failed to read configuration properties",
-                        cb.getCaughtException());
+                method.invokeStaticMethod(C_CREATE_RUN_TIME_CONFIG);
+                method.returnValue(null);
             }
         }
 

--- a/core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java
@@ -188,9 +188,8 @@ public class ApplicationLifecycleManager {
                         }
                         applicationLogger.warn("You can try to kill it with 'kill -9 <pid>'.");
                     }
-                } else if (ExceptionUtil.isAnyCauseInstanceOf(e, ConfigurationException.class)) {
+                } else if (rootCause instanceof ConfigurationException) {
                     System.err.println(rootCause.getMessage());
-                    e.printStackTrace();
                 } else if (rootCause instanceof PreventFurtherStepsException
                         && !StringUtil.isNullOrEmpty(rootCause.getMessage())) {
                     System.err.println(rootCause.getMessage());

--- a/core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java
@@ -24,11 +24,11 @@ import org.wildfly.common.lock.Locks;
 
 import io.quarkus.bootstrap.logging.InitialConfigurator;
 import io.quarkus.bootstrap.runner.RunnerClassLoader;
-import io.quarkus.runtime.configuration.ConfigUtils;
 import io.quarkus.runtime.configuration.ConfigurationException;
 import io.quarkus.runtime.graal.DiagnosticPrinter;
 import io.quarkus.runtime.util.ExceptionUtil;
 import io.quarkus.runtime.util.StringUtil;
+import io.smallrye.config.ConfigValidationException;
 import sun.misc.Signal;
 import sun.misc.SignalHandler;
 
@@ -188,15 +188,13 @@ public class ApplicationLifecycleManager {
                         }
                         applicationLogger.warn("You can try to kill it with 'kill -9 <pid>'.");
                     }
-                } else if (rootCause instanceof ConfigurationException) {
+                } else if (rootCause instanceof ConfigurationException || rootCause instanceof ConfigValidationException) {
                     System.err.println(rootCause.getMessage());
                 } else if (rootCause instanceof PreventFurtherStepsException
                         && !StringUtil.isNullOrEmpty(rootCause.getMessage())) {
                     System.err.println(rootCause.getMessage());
                 } else {
-                    // If it is not a ConfigurationException it should be safe to call ConfigProvider.getConfig here
-                    applicationLogger.errorv(e, "Failed to start application (with profile {0})",
-                            ConfigUtils.getProfiles());
+                    applicationLogger.errorv(e, "Failed to start application");
                     ensureConsoleLogsDrained();
                 }
             }

--- a/core/runtime/src/main/java/io/quarkus/runtime/util/ExceptionUtil.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/util/ExceptionUtil.java
@@ -86,17 +86,6 @@ public class ExceptionUtil {
         return chain.isEmpty() ? null : chain.get(chain.size() - 1);
     }
 
-    public static <T> boolean isAnyCauseInstanceOf(Throwable exception, Class<T> classToCheck) {
-        Throwable curr = exception;
-        do {
-            if (classToCheck.isInstance(curr)) {
-                return true;
-            }
-            curr = curr.getCause();
-        } while (curr != null);
-        return false;
-    }
-
     /**
      * Creates and returns a new {@link Throwable} which has the following characteristics:
      * <ul>

--- a/core/runtime/src/test/java/io/quarkus/runtime/util/ExceptionUtilTest.java
+++ b/core/runtime/src/test/java/io/quarkus/runtime/util/ExceptionUtilTest.java
@@ -13,8 +13,6 @@ import java.util.List;
 
 import org.junit.jupiter.api.Test;
 
-import io.quarkus.runtime.configuration.ConfigurationException;
-
 /**
  *
  */
@@ -61,11 +59,6 @@ public class ExceptionUtilTest {
         assertEquals(NullPointerException.class, ExceptionUtil.getRootCause(new NullPointerException()).getClass());
     }
 
-    @Test
-    public void testIsAnyCauseInstanceOf() {
-        assertTrue(ExceptionUtil.isAnyCauseInstanceOf(generateConfigurationException(), ConfigurationException.class));
-    }
-
     private Throwable generateException() {
         try {
             try {
@@ -86,26 +79,4 @@ public class ExceptionUtilTest {
         }
         throw new RuntimeException("Should not reach here");
     }
-
-    private Throwable generateConfigurationException() {
-        try {
-            try {
-                Integer.parseInt("23.23232");
-            } catch (NumberFormatException nfe) {
-                throw new ConfigurationException("Incorrect param", nfe);
-            }
-        } catch (ConfigurationException ce) {
-            try {
-                throw new IOException("Request processing failed", ce);
-            } catch (IOException e) {
-                try {
-                    throw new IOError(e);
-                } catch (IOError ie) {
-                    return new RuntimeException("Unexpected exception", ie);
-                }
-            }
-        }
-        throw new RuntimeException("Should not reach here");
-    }
-
 }

--- a/extensions/tls-registry/deployment/src/test/java/io/quarkus/tls/PemWithoutKeyTest.java
+++ b/extensions/tls-registry/deployment/src/test/java/io/quarkus/tls/PemWithoutKeyTest.java
@@ -30,7 +30,7 @@ public class PemWithoutKeyTest {
     static final QuarkusUnitTest config = new QuarkusUnitTest().setArchiveProducer(
             () -> ShrinkWrap.create(JavaArchive.class)
                     .add(new StringAsset(configuration), "application.properties"))
-            .assertException(t -> assertThat(t.getCause().getMessage()).contains("quarkus.tls.key-store.pem.0.key"));
+            .assertException(t -> assertThat(t.getMessage()).contains("quarkus.tls.key-store.pem.0.key"));
 
     @Test
     void test() throws KeyStoreException, CertificateParsingException {

